packages/spotlight/src/server/utils/cors.ts

```diff
@@ -18,18 +276,52 @@ export function isAllowedOrigin(origin: string): boolean {
     const url = new URL(origin);
     const hostname = url.hostname.toLowerCase();
 
-    // Allow localhost with any port and protocol (http or https)
-    // Check both [::1] and ::1 formats for IPv6 localhost compatibility
-    if (hostname === "localhost" || hostname === "127.0.0.1" || hostname === "[::1]" || hostname === "::1") {
-      return true;
+    // Fast path: If hostname is already an IP address, check directly
+    if (isIPAddress(hostname)) {
+      return isLocalMachineIP(hostname);
+    }
+
+    // Check cache first
+    const cached = dnsCache.get(hostname);
+    if (cached && cached.expiresAt > Date.now()) {
+      return cached.isLocal;
     }
 
     // Check for spotlightjs.com domains - must be HTTPS with default port
-    if (url.protocol !== "https:" || (url.port !== "" && url.port !== "443")) {
+    // These are always allowed regardless of what they resolve to
+    if (hostname === "spotlightjs.com" || hostname.endsWith(".spotlightjs.com")) {
```

Comment:
This looks like a cheaper check so why not move it earlier than the DNS look ups?